pkg/3scale/amp/component/backend_options.go

```diff
@@ -0,0 +1,98 @@
+package component
+
+import (
+	oprand "github.com/3scale/3scale-operator/pkg/crypto/rand"
+	"github.com/go-playground/validator/v10"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+)
+
+type BackendOptions struct {
```

Comment:
What validation and what defaults would we set for fields in a component that were previously pointers because it allows us to identify not being it set vs being it set with its zero-value?
See for example in the System component the fields *S3FileStorageOptions and *PVCFileStorageOptions. Only one should be set and we need to distinguish between not having provided the option vs having it provided it with the default value of the value it points to.

Reply:
[Validator library](https://godoc.org/github.com/go-playground/validator) is powerful enough for these cases. I made sure it could handle this kind of situations. 

There are struct level validations to validate that either `S3FileStorageOptions` or `PVCFileStorageOptions` fields have been set. See this [example](https://github.com/go-playground/validator/blob/master/_examples/struct-level/main.go#L99)

It also allows conditional validation. So nested struct will be validated when it exists and validation will be omitted when it is nil. For instance:

```
type Address struct {                       
    Street string `validate:"required"`     
    City   string `validate:"required"`   
} 
type User struct { 
    Address        *Address `validate:"omitempty"`
}
```

I think it will be enough to have good validation for the `S3FileStorageOptions` and `PVCFileStorageOptions` fields

Reply:
Ok, if I understand correctly they will still be pointers. They could not be required though because the zero-value of a pointer is nil

Reply:
Yes, optional fields, optional for component package, not for the CR, will be pointers.